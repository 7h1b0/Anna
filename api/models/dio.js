const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const dio = new Schema({
  id_dio: { type: Number, required: true, unique: true },
  name: { type: String, required: true },
});

module.exports = mongoose.model('Dio', dio);
