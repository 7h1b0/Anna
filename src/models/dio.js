const mongoose = require('mongoose');
const Joi = require('joi');

const Schema = mongoose.Schema;

const dio = new Schema({
  id_dio: { type: Number, required: true, unique: true },
  name: { type: String, required: true },
});

dio.statics.validate = function validate(data, callback) {
  const pattern = {
    id_dio: Joi.number()
      .integer()
      .required(),
    name: Joi.string()
      .trim()
      .min(3)
      .required(),
  };

  Joi.validate(data, pattern, callback);
};

dio.statics.pattern = function getPattern() {
  return {
    id_dio: Joi.number()
      .integer()
      .required(),
    name: Joi.string()
      .trim()
      .min(3)
      .required(),
  };
};

module.exports = mongoose.model('Dio', dio);
