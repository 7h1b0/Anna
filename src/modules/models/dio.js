import Ajv from 'ajv';
import dioSchema from '../schemas/dio';
import knex from '../../knexClient';
import { returnFirst } from '../dbUtil';

export const TABLE = 'dios';

export function validate(data) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export function findAll() {
  return knex(TABLE).select({ dioId: 'dio_id' }, { roomId: 'room_id' }, 'name');
}

export function findById(id) {
  return returnFirst(
    knex(TABLE)
      .select({ dioId: 'dio_id' }, { roomId: 'room_id' }, 'name')
      .where('dio_id', id),
  );
}

export function save({ dioId, roomId, name }) {
  return knex(TABLE)
    .insert({ dio_id: dioId, room_id: roomId, name })
    .then(() => {
      return { dioId, roomId, name };
    });
}

export function remove(dioId) {
  return knex(TABLE)
    .where('dio_id', dioId)
    .del();
}

export function findByIdAndUpdate(dioId, payload) {
  return knex(TABLE)
    .update(payload)
    .where('dio_id', dioId);
}
