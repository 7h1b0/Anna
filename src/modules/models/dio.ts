import Ajv from 'ajv';
import dioSchema from '../schemas/dio';
import knex from '../../knexClient';

export const TABLE = 'dios';

export interface Dio {
  dioId: number;
  roomId: string;
  name: string;
}

export function validate(data: object) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export async function findAll(): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name');
}

export async function findById(dioId: number): Promise<Dio> {
  return knex(TABLE)
    .first('dioId', 'roomId', 'name')
    .where('dioId', dioId);
}

export async function save(newDio: Dio): Promise<Dio> {
  return knex(TABLE)
    .insert(newDio)
    .then(() => newDio);
}

export function remove(dioId: number) {
  return knex(TABLE)
    .where('dioId', dioId)
    .del();
}

export function findByIdAndUpdate(dioId: number, payload: Partial<Dio>) {
  const { dioId: ignored, ...dio } = payload;
  return knex(TABLE)
    .update(dio)
    .where('dioId', dioId);
}
