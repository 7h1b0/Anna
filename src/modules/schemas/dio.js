module.exports = {
  $id: 'dio',
  type: 'object',
  additionalProperties: false,
  properties: {
    dioId: {
      type: 'integer',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    roomId: {
      type: 'integer',
    },
  },
  required: ['name', 'dioId', 'roomId'],
};
