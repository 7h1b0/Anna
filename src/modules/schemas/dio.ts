export default {
  $id: 'dio',
  type: 'object',
  additionalProperties: false,
  properties: {
    dioId: {
      type: 'integer',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    roomId: {
      type: 'string',
      format: 'uuid',
    },
  },
  required: ['name', 'dioId', 'roomId'],
};
