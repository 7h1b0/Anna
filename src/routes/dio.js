const Dio = require('../models/dio');
const { getJoiError, actions, dispatch } = require('../utils/');

module.exports = (app) => {
  app.route('/api/dios')
    .get((req, res) => {
      Dio.find({})
        .then(dios => res.send(dios))
        .catch(err => res.status(500).send({ err }));
    })

    .post((req, res) => {
      Dio.validate(req.body, (invalidReq, dio) => {
        if (invalidReq) {
          res.status(400).send(getJoiError(invalidReq));
        } else {
          new Dio(dio).save()
            .then(newDio => res.status(201).send(newDio))
            .catch(err => res.status(500).send({ err }));
        }
      });
    });

  app.route('/api/dios/:id_dio([0-9]{1,2})')
    .get((req, res) => {
      Dio.findOne({ id_dio: req.params.id_dio })
        .then((dio) => {
          if (!dio) {
            res.sendStatus(404);
          } else {
            res.send(dio);
          }
        })
        .catch(err => res.status(500).send({ err }));
    })

    .put((req, res) => {
      Dio.validate(req.body, (invalidReq, dio) => {
        if (invalidReq) {
          res.status(400).send(getJoiError(invalidReq));
        } else {
          Dio.findOneAndUpdate({ id_dio: req.params.id_dio }, dio, { new: true })
            .then((dio) => {
              if (!dio) {
                res.sendStatus(404);
              } else {
                res.send(dio);
              }
            })
            .catch(err => res.status(500).send({ err }));
        }
      });
    })

    .delete((req, res) => {
      Dio.findOneAndRemove({ id_dio: req.params.id_dio })
        .then((dio) => {
          if (!dio) {
            res.sendStatus(404);
          } else {
            res.sendStatus(204);
          }
        })
        .catch(err => res.status(500).send({ err }));
    });


  app.get('/api/dios/:id_dio([0-9]{1,2})/:status(on|off)', (req, res) => {
    dispatch(actions.toggleDio(req.params.id_dio, req.params.status === 'on'));
    res.end();
  });
};
