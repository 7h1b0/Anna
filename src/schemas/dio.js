module.exports = {
  $id: 'dio',
  type: 'object',
  additionalProperties: false,
  properties: {
    id_dio: {
      type: 'integer',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
  },
  required: ['name', 'id_dio'],
};
