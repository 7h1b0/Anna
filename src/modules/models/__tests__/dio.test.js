import knex from '../../../knexClient';
import * as Dio from '../dio';
const initDios = [
  {
    dio_id: 1,
    room_id: 1,
    name: 'test',
  },
  {
    dio_id: 2,
    room_id: 2,
    name: 'test',
  },
];

describe('Dio', () => {
  beforeAll(async () => {
    await knex(Dio.TABLE).truncate();
  });

  beforeEach(async () => {
    await knex(Dio.TABLE).insert(initDios);
  });

  afterEach(async () => {
    await knex(Dio.TABLE).truncate();
  });

  afterAll(async () => {
    await knex.destroy();
  });

  describe('findAll', () => {
    it('should return all dios', async () => {
      const expected = [
        {
          dioId: 1,
          roomId: 1,
          name: 'test',
        },
        {
          dioId: 2,
          roomId: 2,
          name: 'test',
        },
      ];

      const result = await Dio.findAll();
      expect(result).toEqual(expected);
    });
  });

  describe('findById', () => {
    it('should return only one dio', async () => {
      const expected = {
        dioId: 1,
        roomId: 1,
        name: 'test',
      };

      const result = await Dio.findById(1);
      expect(result).toEqual(expected);
    });

    it('should return undefined', async () => {
      const result = await Dio.findById(-1);
      expect(result).toBe(undefined);
    });
  });

  describe('save', () => {
    it('should save a new dio', async () => {
      const save = {
        dioId: 3,
        roomId: 1,
        name: 'test-save',
      };

      const newDio = await Dio.save(save);
      expect(newDio).toEqual(save);
      const dios = await knex(Dio.TABLE)
        .select('*')
        .where('dio_id', 3);
      expect(dios).toMatchSnapshot();
    });

    it('should reject when an id is already taken', async () => {
      const save = {
        dioId: 2,
        roomId: 1,
        name: 'test-save',
      };

      await expect(Dio.save(save)).rejects.toBeDefined();
    });
  });

  describe('delete', () => {
    it('should delete a dio', async () => {
      await Dio.remove(1);
      const dios = await knex(Dio.TABLE).select('*');
      expect(dios).toMatchSnapshot();
    });

    it('should not delete a dio', async () => {
      await Dio.remove(-1);
      const dios = await knex(Dio.TABLE).select('*');
      expect(dios).toEqual(initDios);
    });
  });

  describe('findByIdAndUpdate', () => {
    it('should update a dio', async () => {
      const rowsAffected = await Dio.findByIdAndUpdate(1, { name: 'updated' });
      expect(rowsAffected).toBe(1);
      const dios = await knex(Dio.TABLE).select('*');
      expect(dios).toMatchSnapshot();
    });

    it('should not update a dio', async () => {
      const rowsAffected = await Dio.findByIdAndUpdate(-1, { name: 'updated' });
      expect(rowsAffected).toBe(0);
      const dios = await knex(Dio.TABLE).select('*');
      expect(dios).toEqual(initDios);
    });
  });

  describe('validate', () => {
    it('should return true when a dio is valid', () => {
      const dio = {
        dioId: 3,
        roomId: 1,
        name: 'test-save',
      };

      expect(Dio.validate(dio)).toBeTruthy();
    });

    it('should return false when a dio is missing a props', () => {
      const dio = {
        dioId: 3,
        name: 'test-save',
      };

      expect(Dio.validate(dio)).toBeFalsy();
    });

    it('should return false when a dio has unknow props', () => {
      const dio = {
        dioId: 3,
        roomId: 1,
        name: 'test-save',
        test: 2,
      };

      expect(Dio.validate(dio)).toBeFalsy();
    });
  });
});
