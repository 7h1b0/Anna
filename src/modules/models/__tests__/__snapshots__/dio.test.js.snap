// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dio findAll should return all dios 1`] = `
Array [
  RowDataPacket {
    "dioId": 1,
    "name": "test",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  },
  RowDataPacket {
    "dioId": 2,
    "name": "test",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  },
]
`;

exports[`Dio findById should return only one dio 1`] = `
RowDataPacket {
  "dioId": 1,
  "name": "test",
  "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
}
`;

exports[`Dio findByIdAndUpdate should update a dio 1`] = `
Array [
  RowDataPacket {
    "dioId": 1,
    "name": "updated",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  },
  RowDataPacket {
    "dioId": 2,
    "name": "test",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  },
]
`;
